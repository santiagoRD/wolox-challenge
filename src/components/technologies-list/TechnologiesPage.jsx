import React, { useEffect, lazy, Suspense, useMemo, useState } from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { userLogout } from '../../actions/authActions';
import { getTechsByNameAndType, noop, sortDesc, sortAsc } from '../../helpers/helpers';
import { BASE_URL } from '../../helpers/constants';
import { setTechnologies, techError } from '../../actions/technologiesActions';
import Loader from '../../helpers/common/Loader';
import './TechnologiesList.scss';

const TechnologiesNav = lazy(() => import('./TechnologiesNav'));
const TechnologiesFooter = lazy(() => import('./TechnologiesFooter'));
const TechnologiesList = lazy(() => import('./TechnologiesList'));

const TechnologiesPage = props => {
  const [techList, setTechList] = useState([]);
  const { handleSetTechs, handleSetError, handleLogout } = props;
  const [searchText, setSearchText] = useState('');
  const [updateData, setUpdateData] = useState(true);
  const filteredTechList = useMemo(() => getTechsByNameAndType(techList, searchText), [techList, searchText, setTechList]);

  const handleUserLogout = () => {
    localStorage.removeItem('auth');
    handleLogout();
  };

  const handleInputChange = ({ target }) => {
    setSearchText(target.value);
  };

  const handleUpdateData = techs => {
    localStorage.setItem('techList', JSON.stringify(techs));
    setUpdateData(true);
  };

  const handleSelectFavorite = name => {
    const tmpTechList = techList.map(tech => (tech.tech === name ? { ...tech, favorite: !tech.favorite } : tech));
    handleUpdateData(tmpTechList);
  };

  const getTechList = async () => {
    try {
      let { data } = await axios.get(`${BASE_URL}/techs`);
      data = data.map(tech => ({ ...tech, favorite: false }));
      setTechList(data);
      handleSetTechs(data);
    } catch (error) {
      handleSetError(`Hubo un error obteniendo el listado, ${error}`);
    }
  };

  const handleCalculateFavorites = () => {
    let quantity = 0;
    techList.forEach(tech => {
      if (tech.favorite) quantity += 1;
    });
    return quantity;
  };

  const handleSortAsc = () => {
    const sortedTechs = sortAsc(techList);
    handleUpdateData(sortedTechs);
  };
  const handleSortDesc = () => {
    const sortedTechs = sortDesc(techList);
    handleUpdateData(sortedTechs);
  };

  const favorites = useMemo(() => handleCalculateFavorites(), [techList]);

  useEffect(() => {
    if (updateData) {
      if (localStorage.getItem('techList')) {
        setTechList(JSON.parse(localStorage.getItem('techList')));
        handleSetTechs(JSON.parse(localStorage.getItem('techList')));
      } else {
        getTechList();
      }
      setUpdateData(false);
    }
  }, [updateData]);
  return (
    <section className='techlist__section'>
      <Suspense fallback={<Loader />}>
        <TechnologiesNav
          searchText={searchText}
          handleInputChange={handleInputChange}
          favorites={favorites}
          handleSortAsc={handleSortAsc}
          handleSortDesc={handleSortDesc}
          handleUserLogout={handleUserLogout}
        />
        <TechnologiesList techs={filteredTechList} handleSelectFavorite={handleSelectFavorite} />
        <TechnologiesFooter total={filteredTechList.length} />
      </Suspense>
    </section>
  );
};

const mapDispatchToProps = dispatch => bindActionCreators({ handleLogout: userLogout, handleSetTechs: setTechnologies, handleSetError: techError }, dispatch);

export default connect(noop, mapDispatchToProps)(TechnologiesPage);

TechnologiesPage.propTypes = {
  handleSetTechs: PropTypes.func.isRequired,
  handleSetError: PropTypes.func.isRequired,
  handleLogout: PropTypes.func.isRequired
};
