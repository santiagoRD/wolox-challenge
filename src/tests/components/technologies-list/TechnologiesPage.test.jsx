import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { MemoryRouter } from 'react-router-dom';
import TechnologiesPage from '../../../components/technologies-list/TechnologiesPage';

const middlewares = [];
const mockStore = configureStore(middlewares);

const initialState = {
  techs: [],
  error: ''
};
let store = mockStore(initialState);
const wrapper = mount(
  <MemoryRouter>
    <Provider store={store}>
      <TechnologiesPage />
    </Provider>
  </MemoryRouter>
);

describe('Testing <TechnologiesPage />', () => {
  beforeEach(() => {
    store = mockStore(initialState);
  });
  test('should render <TechnologiesPage />', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
