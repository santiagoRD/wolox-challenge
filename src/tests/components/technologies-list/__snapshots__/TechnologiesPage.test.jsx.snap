// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <TechnologiesPage /> should render <TechnologiesPage /> 1`] = `
<section
  className="techlist__section"
>
  <Suspense
    fallback={<Loader />}
  >
    <div
      className="loader__wrapper"
    >
      <div
        className="loader__container"
      >
        <div
          className="loader__item"
        />
      </div>
    </div>
  </Suspense>
</section>
`;
